### PR TITLE
Make case-insensitive class lookup work correctly with exported types

### LINKED ISSUE #15153: [netcore] Make System.Tests.ActivatorTests.TestingCreateInstanceObjectHandleFullSignature Pass
The test currently fails with:

```
System.Tests.ActivatorTests.TestingCreateInstanceObjectHandleFullSignature(assemblyName: "mscorlib", type: "System.Nullable`1[[System.Int32, mscorlib, Version"..., ignoreCase: True, bindingAttr: Instance | Public, binder: DefaultBinder { }, args: [], culture: , activationAttributes: null, returnedFullNameType: "", returnNull: True) [FAIL]
      System.TypeLoadException : Could not load type 'System.Nullable`1[[System.Int32, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]]' from assembly 'mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'.
      Stack Trace:
           at System.Reflection.RuntimeAssembly.GetType(String name, Boolean throwOnError, Boolean ignoreCase)
           at System.Activator.CreateInstanceInternal(String assemblyString, String typeName, Boolean ignoreCase, BindingFlags bindingAttr, Binder binder, Object[] args, CultureInfo culture, Object[] activationAttributes, StackCrawlMark& stackMark)
           at System.Activator.CreateInstance(String assemblyName, String typeName, Boolean ignoreCase, BindingFlags bindingAttr, Binder binder, Object[] args, CultureInfo culture, Object[] activationAttributes)
           at System.Tests.ActivatorTests.TestingCreateInstanceObjectHandleFullSignature(String assemblyName, String type, Boolean ignoreCase, BindingFlags bindingAttr, Binder binder, Object[] args, CultureInfo culture, Object[] activationAttributes, String returnedFullNameType, Boolean returnNull)
           at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
```

Determine why and make the test pass.

### PR DESCRIPTION
**Pull request by CoffeeFlux**
Fixes #15153 

This changes `mono_class_from_name_case_checked` to call into `mono_class_from_name_checked_aux`, which I made case-aware. The old code was broken in a couple different ways and almost never hit, as the only way to do so from C# was via reflection.

This is neither terribly pretty nor fast, but hopefully it sees little use outside of the test suite.

### CODE REVIEW

**Comment by CoffeeFlux** [comment]
The issue I'm hitting with this PR is that when you use the cache, you have to check *all* matching namespace caches rather than just the first. I've pushed a commit to fix that, but it runs into an important issue: what order do you search the namespaces in? 

This is somewhat of a mess with all the case-insensitive lookup, and the CoreFX test suite recognizes it as such and only ensures we return a *valid* match, not some specific one as that's very heavily implementation dependent. This implementation, then, should pass all the tests.

However, it leads to use having different behavior from Core and in general somewhat strange selection behavior dependent on the hash function. For example, [this snippet](https://gist.github.com/CoffeeFlux/c8a4f584caf32163bfe0fb6578be512e) returns the following:
```
HelloWorld.MyNamespace1.MynAmespace3.gOo`1[System.Int32]
```

The alternative is to fall back to reiterating over the typedef and exported types tables, which is slower but will not change the order we iterate in. I've implemented that as well and pushed it to a branch [here](https://github.com/CoffeeFlux/mono/tree/case-insensitive-names-alt). When running the above sample, it returns the following:
```
HelloWorld.MyNamespace1.MynAmespace3.Goo`1[System.Int32]
```

I think with the inherent ambiguity of the case-insensitive lookup, the cache-based implementation is fine and significantly faster, but I've written this in an effort to document that decision should someone (including myself) run into compatibility-related issues down the line.

**Comment by CoffeeFlux** [comment]
@monojenkins build failed

**Comment by CoffeeFlux** [comment]
Remaining failure occurs when you build the entire project (make -j12 in the git root), but if you rebuild the runtime itself it goes away. It's probably invalidating some AOT'd image and the solution is to fix `get_class_from_name` to be case-insensitive as well, but I'll have to test.

**Comment by CoffeeFlux** [comment]
@monojenkins build failed

**Comment by CoffeeFlux** [comment]
Oh, forgot to link it back but I also pr'd corefx with a test addition that covers one of the scenarios fixed here: https://github.com/dotnet/corefx/pull/40788

**Comment by kg** [comment]
@monojenkins build failed

**Comment by CoffeeFlux** [comment]
Test failure might be relevant? I haven't the faintest idea why though, since the AOT compiler shouldn't be using case-insensitive lookup.

`Unable to compile method 'void MyType:Method ()' due to: 'Could not load file or assembly 'Dep, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null' or one of its dependencies.'.`

**Comment by CoffeeFlux** [comment]
@monojenkins build failed

File: mono/metadata/class.c [hunk 1]
```diff
@@ -3157,10 +3140,32 @@ mono_class_from_name_checked_aux (MonoImage *image, const char* name_space, cons
...
+			token = GPOINTER_TO_UINT (g_hash_table_lookup (nspace_table, name));
+	} else {
+		FindAllUserData all_user_data = { name_space, NULL };
+		FindUserData user_data = { name, NULL };
+		GSList *values;
 
-	if (nspace_table)
-		token = GPOINTER_TO_UINT (g_hash_table_lookup (nspace_table, name));
+		// We're forced to check all matching namespaces, not just the first one found,
+		// because our desired type could be in any of the ones that match case-insensitively.
```
Comment (lambdageek) [inline mono/metadata/class.c]: Wow

**Pull request merged**